File: csharp-workshop/Part3.MocksClean/Service/OrderCancellationService.cs
using Part3.MocksClean.Domain;
using Part3.MocksClean.Repository;

namespace Part3.MocksClean.Service;

public class OrderCancellationService
{
    private readonly IOrderRepository _orderRepository;
    private readonly IPaymentService _paymentService;
    private readonly IStockService _stockService;
    private readonly IShippingService _shippingService;

    public OrderCancellationService(
        IOrderRepository orderRepository,
        IPaymentService paymentService,
        IStockService stockService,
        IShippingService shippingService)
    {
        _orderRepository = orderRepository;
        _paymentService = paymentService;
        _stockService = stockService;
        _shippingService = shippingService;
    }

    public CancellationResult CancelOrder(string orderId)
    {
        var order = _orderRepository.FindById(orderId);
        if (order == null)
        {
            return new CancellationResult.Failure("Order not found");
        }

        if (order.Status != OrderStatus.Confirmed)
        {
            return new CancellationResult.Failure("Order cannot be cancelled");
        }

        if (order.TransactionId != null)
        {
            // THE TRAP: treats false as ERROR, but PaymentService means "nothing to refund"
            if (!_paymentService.RefundPayment(order.TransactionId))
            {
                return new CancellationResult.Failure("Refund failed");
            }
        }

        if (order.TrackingNumber != null && !_shippingService.CancelShipment(order.TrackingNumber))
        {
            return new CancellationResult.Failure("Shipment cancellation failed");
        }

        var reservations = order.Items
            .Select(item => new StockReservation(item.ProductId, item.Quantity, true))
            .ToList();
        _stockService.ReleaseStock(reservations);
        _orderRepository.UpdateStatus(orderId, OrderStatus.Cancelled);
        return new CancellationResult.Success();
    }
}

File: typescript-workshop/part-4-behavior/src/service/OrderCancellationService.ts
import { OrderStatus } from '../domain/OrderStatus';
import { StockReservation } from '../domain/StockReservation';
import { CancellationResult } from '../domain/CancellationResult';
import { OrderRepository } from '../repository/OrderRepository';
import { PaymentService } from './PaymentService';
import { ShippingService } from './ShippingService';
import { StockService } from './StockService';

export class OrderCancellationService {
  constructor(
    private readonly orderRepository: OrderRepository,
    private readonly paymentService: PaymentService,
    private readonly stockService: StockService,
    private readonly shippingService: ShippingService,
  ) {}

  cancelOrder(orderId: string): CancellationResult {
    const order = this.orderRepository.findById(orderId);
    if (order === null) {
      return CancellationResult.failure('Order not found');
    }

    if (order.transactionId !== null) {
      const refunded = this.paymentService.refundPayment(order.transactionId);
      if (!refunded) {
        return CancellationResult.failure('Refund failed');
      }
    }

    const reservations = order.items.map(
      (item) => new StockReservation(item.productId, item.quantity, true),
    );
    this.stockService.releaseStock(reservations);

    if (order.trackingNumber !== null) {
      this.shippingService.cancelShipment(order.trackingNumber);
    }

    this.orderRepository.updateStatus(orderId, OrderStatus.Cancelled);
    return CancellationResult.success();
  }
}
